dlib/cuda/cudnn_dlibapi.cpp

```diff
@@ -170,6 +170,18 @@ namespace dlib
             return des.get_handle();
         }
 
+        static cudnnActivationDescriptor_t clipped_relu_activation_descriptor(float coef)
+        {
+            thread_local cudnn_activation_descriptor des(CUDNN_ACTIVATION_CLIPPED_RELU, CUDNN_PROPAGATE_NAN, coef);
```

Comment:
So this doesn't work because `coef` is used only on the first call to `clipped_relu_activation_descriptor()`.  All subsequent calls (from the same thread anyway) are going to ignore whatever `coef` is given.  That's probably why you are seeing the odd behavior you noted.

Reply:
Anyway, I don't know what costs there are in descriptor creation in cudnn.  The API is pretty suggestive of there at least being a dynamic allocation, which sucks.  But maybe it's not doing that and it's just a stack local variable through and through.  Maybe just moving `cudnnSetActivationDescriptor` calls out so you can set `coef` is fine.  You should check if there is any notable runtime cost to calling those cudnn functions that we currently avoid calling more than one via use of `thread_local`.

Reply:
OK, I will test it soon (if I understood it correctly, thanks for looking into it :))

Honestly, I don't need these activations myself, but I thought it was a shame that they were already implemented in CUDNN and not in dlib. This was supposed to be a simple PR, haha.

Reply:
This can't be `thread_local` since only the first `coeff` will be used.   Like if you add a unit test that tests `clipped_relu()` that first calls it with a small coeff and then with a bigger one you should see the second call return the wrong result.